File: 53-maximum-subarray/maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int current_sum = 0;
        int maxi = INT_MIN;
        
        for (int num : nums) {
            current_sum = max(num, current_sum + num);  // Take the current number or add to the previous sum
            maxi = max(maxi, current_sum);  // Update the maximum sum
        }
        
        return maxi;
    }
};

File: 53-maximum-subarray/maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # Optimal
        
        # maxi = float("-inf")

        # for i in range(0, len(nums)):
        #     total = 0
        #     for j in range(i, len(nums)):
        #         total = total + nums[j]
        #         maxi = max(maxi, total)
        
        # return maxi

        maxi = float("-inf")
        total = 0

        for i in range(0, len(nums)):
            total = total + nums[i]
            maxi = max(maxi, total)

            if total < 0:
                total = 0

        return maxi

